src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.hh

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef IGNITION_GAZEBO_SYSTEMS_LOGICAL_AUDIO_SENSOR_PLUGIN_HH_
+#define IGNITION_GAZEBO_SYSTEMS_LOGICAL_AUDIO_SENSOR_PLUGIN_HH_
+
+#include <memory>
+
+#include <ignition/gazebo/System.hh>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace systems
+{
+  // Forward Declaration
+  class LogicalAudioSensorPluginPrivate;
+
+  /// \brief A plugin for trivial audio detection.
+  ///
+  /// Each <plugin> tag can only accept one sensor (either a sound source
+  /// or microphone).
+  /// In order to create multiple audio sources and microphones,
+  /// multiple <plugin> tags must be used (one for each sensor).
+  /// After each simulation step, microphones check if audio
+  /// was detected by any sources in the world.
+  /// No audio is actually played to an audio device
+  /// such as speakers - this plugin is meant to check if audio
+  /// could theoretically be heard at a certain location or not.
+  ///
+  /// Secifying an audio source via SDF is done as follows:
+  ///
+  /// <source> A new audiosource in the environment, which has the
+  ///   following child elements:
+  ///     * <id> The source ID, which must be unique and >= 0.
```

Comment:
I have three questions:

1. Unique between all audio sources in the world? If so, could you add some language to that effect?
2. Also, what happens if two audio sources with the same ID are added?
3. Can the audio id be created automatically at run time? This would eliminate this xml attribute and guarantee that all audio source IDs are unique.

Reply:
My response here applies to both the microphone and source IDs:

> Unique between all audio sources in the world? If so, could you add some language to that effect?

Yes, that's correct - I'll add some clarifying language here.

> Also, what happens if two audio sources with the same ID are added?

If a source has the same ID as another source, this source won't be created since a source with this ID already exists (the same is true for a microphone) - here's the [source ID checking code](https://github.com/ignitionrobotics/ign-gazebo/blob/7fc021315c855c28d0076c1593d50a7667bff00c/src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc#L280-L296), and here's the [microphone ID checking code](https://github.com/ignitionrobotics/ign-gazebo/blob/7fc021315c855c28d0076c1593d50a7667bff00c/src/systems/logical_audio_sensor_plugin/LogicalAudioSensorPlugin.cc#L431-L447). Both of these checks occur in the system plugin's `configure` method.

As you can see in the code links I just shared, the sources/microphones with duplicate IDs aren't created, but the simulation still runs (the simulation just runs without the duplicate sources/microphones) - would it be better to end simulation early/not run the simulation at all so that it's clear to users that something was not configured correctly? Now that I am taking another look at this, my fear is that users will not realize one of their sources/microphones wasn't initialized if we still run the simulation after detecting duplicate sources/microphones.

> Can the audio id be created automatically at run time? This would eliminate this xml attribute and guarantee that all audio source IDs are unique.

Technically, yes the IDs can be created at run time. I don't think that this is a good idea though, because this would mean that the users have no idea which topics/services are associated with which sources/microphones. Right now, the IDs provide a way for users to know how to interact with specific microphones/sources in the environment since an audio object's ID is embedded in the topic/service name (for example, by giving a source an ID of 5 in the SDF file, I now know that if I want to perform an action on this source, I just need to call a service with `source_5` in it). If IDs were automatically generated, then I'm not sure how we would create topics/services for sources/microphones in a way that would allow users to know which sources/microphones they are interacting with.

I don't think that it's too much to ask for users to give sources/microphones an ID attribute that's unique in the SDF, as it's actually making things easier for them since they now have a way to interact with every source/microphone directly. However, if this is not a good approach, I am open to other ideas.

Reply:
After getting some other feedback, I'll move towards an approach of removing IDs from audio objects and pre-pending an audio object's parent model name to its associated service/topic. I'll work on that and let you know when the changes are ready.

Reply:
I made a change so that the topics/services are in the parent entity's namespace (they used to be in the global namespace), along with enforcing ID uniqueness at the parent entity level instead of globally in https://github.com/ignitionrobotics/ign-gazebo/pull/401/commits/bf5cacbbe9aa1d37806a99d83cd6d9c5e56ef1c4.